New home screen and layout changes

kleinMaggus opened the pull request:
I have thought about how to visualize the documentations in the app and decided to split up the work into the layout changes including a new home screen. It's an idea and we can discuss it. No problem if you don't like it ;)

First I created a new home screen:
![Screenshot_Home](https://user-images.githubusercontent.com/3748388/117012049-ceffda00-acee-11eb-8e48-b500421b142f.png)
![Screenshot_Home_All](https://user-images.githubusercontent.com/3748388/117012061-d2936100-acee-11eb-835a-5411fd4b16aa.png)

My idea behind this Disney+ inspired screen is that the user mainly wants to look a stream of the current season. So I created the big teaser image which should be focussed first and looks quite good and prominent. After that I created the visualization of the "Archive" with the last 5 season and a option to display all seasons. It's quite the "Category"-Thinking of all big streaming applications and I would like to display the "Documentations" below the Archive category.

After that I had a look at the F1TV website and saw that the seasons in the archive were structured in decades. So I switched the layout to the BrowseSupportFragment.
![Screenshot_Season_Selection](https://user-images.githubusercontent.com/3748388/117012807-97456200-acef-11eb-91af-793943d0848a.png)
Here I'm not sure whether to display the category title as "2009-2000" or like I did "2000-2009".

And at the end I playing around with some colors (which I currently not have commited) to make it more F1 like. If you like it I can give it a commit/push ;)
![Screenshot_Red](https://user-images.githubusercontent.com/3748388/117013046-cc51b480-acef-11eb-9eb9-0f36729279b2.png)

kleinMaggus commented:
And I need some dutch translations for this @bashopman if we are going to merge it :D

bashopman commented on app/src/main/res/values/strings.xml:
```diff
@@ -22,4 +22,11 @@
 
     <string name="choose_a_season">Select a season</string>
+
+    <string name="teaser_image_text">Current season: %1$d</string>
```
```suggestion
    <string name="teaser_image_text">Huidig seizoen: %1$d</string>
```

bashopman commented on app/src/main/res/values/strings.xml:
```diff
@@ -22,4 +22,11 @@
     <string name="choose_a_season">Select a season</string>
+
+    <string name="teaser_image_text">Current season: %1$d</string>
+    <string name="teaser_image_description">Current season</string>
```
```suggestion
    <string name="teaser_image_description">Huidig seizoen</string>
```

bashopman commented on app/src/main/res/values/strings.xml:
```diff
@@ -22,4 +22,11 @@
+    <string name="teaser_image_text">Current season: %1$d</string>
+    <string name="teaser_image_description">Current season</string>
+
+    <string name="home_all">Show all</string>
```
```suggestion
    <string name="home_all">Toon  alle</string>
```

bashopman commented on app/src/main/res/values/strings.xml:
```diff
@@ -22,4 +22,11 @@
+    <string name="teaser_image_description">Current season</string>
+
+    <string name="home_all">Show all</string>
+    <string name="home_archive">Archive</string>
```
```suggestion
    <string name="home_archive">Archief</string>
```

bashopman reviewed (commented):
Hi kleinMaggus,

Thanks for inviting me! I have added the translations as suggestions. Obviously you'll have to copy those in the Dutch `strings.xml` file... ;)


From a non Android developer, I'm impressed with what you built. Nice.


As an app user, I'm not sure about the added value of the extra home-screen. 

From the screenshots it is not immediately clear for me that the teaser image is clickable. I don't know what it looks like on Disney+, but maybe you can accentuate that a bit more?

Additionally, the string "current season: 2021" over the teaser image could be just "2021" for me and be just as clear.


I'm really interested to see what @leonardoxh  thinks.

Regards,

    Bas

kleinMaggus commented:
Hi Bas,

> Thanks for inviting me! I have added the translations as suggestions. Obviously you'll have to copy those in the Dutch strings.xml file... ;)

Thanks for the translations. I will add them asap :)

> From a non Android developer, I'm impressed with what you built. Nice.

Well thanks man. I'm also not an Android Developer too and very happy about being a Web Developer, so much more flexibilities and possibilities :D

> As an app user, I'm not sure about the added value of the extra home-screen.

I wanted to include the documentaries and was thinking about the navigation in the app and thought an additional home screen would be nice as it is known in other apps (Netflix, Prime, Disney+). Sure, for now the value is not that big because I implemented the screen without extra categories (Documentaries below the archive, maybe the current/last race above the archive), but I wanted to discuss about the layout/navigation first before starting and creating a big PR.

> From the screenshots it is not immediately clear for me that the teaser image is clickable. I don't know what it looks like on Disney+, but maybe you can accentuate that a bit more?

Ok, I was trying to clarify that with the white border around the image and the white border around the archive items. Maybe we can increase the border width? Or create some "zoom"-effect on the image? But thats beyond my android knowledge :D

> Additionally, the string "current season: 2021" over the teaser image could be just "2021" for me and be just as clear.

Yeah, that was what I was thinking too. But I wanted some other opinions first :)

Hope I could clarify some things? :)

bashopman commented:
Hi kleinMaggus,

With the additional plans you have for the home screen, it makes much more sense! Thanks for explaining. I particularly like the idea of showing the current/last broadcasts!

Maybe you can make the design of the final concept and then gradually work towards it in a few pull requests?

Regards,

   Bas

leonardoxh commented:
Hi @kleinMaggus hehe nice stuff man, 

I couldn't run the project yet due some work related stuff, but what is the current implemented design ? 

The one I like most is 2nd screen (show all) and the 3rd screen (the blue background), I am not really sure if we should use these red colours as far as they are nice with the F1 brand itself, that is the main problem, we can't really refer their brand a lot (or I might be sued).

I liked that you left a gap for documentaries which is something that I will probably implement on future.

Can we hammer down on the design with 2nd and 3rd screen? let me know on what you think, but I liked the design a lot, thanks for the effort there.

kleinMaggus commented:
Hi @leonardoxh 

the first 3 screens are currently implemented. The 1st one shows the focus state of the image (which you can click on) and the list row when you open the app. The 2nd screens shows that you can navigate to the list row and to the right (to show all). And the 3rd one is using the blue background like in the session selection (when you have selected the current season for example). 

With the red background I'm not sure too. Just wanted to show how it can look like. The red color is one of our banner logo, so it's not the red from F1 :D But I think we will go with the blue one as we can adjust colors later :)

leonardoxh commented on app/build.gradle.kts:
```diff
@@ -14,8 +14,8 @@ android {
-        versionCode = 21
-        versionName = "2.0.7"
+        versionCode = 22
+        versionName = "2.0.8"
```
I think this change would qualify for 2.1.0 🗡️

leonardoxh commented on app/src/main/java/fr/groggy/racecontrol/tv/ui/home/HomeFragment.kt:
```diff
@@ -0,0 +1,115 @@
+    ): View {
+        val view = super.onCreateView(inflater, container, savedInstanceState)
+        val params = view!!.layoutParams as ViewGroup.MarginLayoutParams
+        val resources: Resources = inflater.context.resources
+
+        val dimensionPixelSize = resources.getDimensionPixelSize(R.dimen.lb_browse_rows_fading_edge)
```
Just use these `resources` directly without add it to a variable...

Like...
inflater.context.resources.resources.getDimensionPixelSize

leonardoxh commented on app/src/main/java/fr/groggy/racecontrol/tv/ui/home/HomeFragment.kt:
```diff
@@ -0,0 +1,115 @@
+        val dimensionPixelSize = resources.getDimensionPixelSize(R.dimen.lb_browse_rows_fading_edge)
+        val horizontalMargin = -dimensionPixelSize * 2 - 4
+
+        params.leftMargin = horizontalMargin
+        params.rightMargin = horizontalMargin
+        view.layoutParams = params
+        return view
```
With these layout params, there is a nice ktx extension for it:

use the view?.updateLayoutParams<ViewGroup.MargingLayoutParams> { /* your update code */ }

leonardoxh replied:
btw, I think the margingLayoutParams has a `horizontalMargin` property

kleinMaggus replied:
I updated the code to use updateLayoutParams. But I didn't find any `horizontalMargin` option.

leonardoxh commented on app/src/main/java/fr/groggy/racecontrol/tv/ui/home/HomeItem.kt:
```diff
@@ -0,0 +1,10 @@
+package fr.groggy.racecontrol.tv.ui.home
+
+enum class HomeItemType{
```
whitespace before `{`

leonardoxh commented on app/src/main/java/fr/groggy/racecontrol/tv/ui/season/archive/SeasonArchiveFragment.kt:
```diff
@@ -13,28 +13,43 @@ import fr.groggy.racecontrol.tv.ui.season.browse.SeasonBrowseActivity
         val archives = viewModel.listArchive()
-        itemAdapter.setItems(archives, null)
+        val decades = archives.groupBy { archive: Archive -> (archive.year / 10) * 10 }
+        decades.entries.forEach {decade ->
```
whitespace after `{`

leonardoxh commented on app/src/main/java/fr/groggy/racecontrol/tv/ui/season/archive/SeasonArchiveFragment.kt:
```diff
@@ -13,28 +13,43 @@ import fr.groggy.racecontrol.tv.ui.season.browse.SeasonBrowseActivity
-        itemAdapter.setItems(archives, null)
+        val decades = archives.groupBy { archive: Archive -> (archive.year / 10) * 10 }
+        decades.entries.forEach {decade ->
+            run {
```
`run` here isn't really necessary

leonardoxh commented on app/src/main/java/fr/groggy/racecontrol/tv/ui/home/HomeItemCardView.kt:
```diff
@@ -0,0 +1,32 @@
+    private val textView: TextView
+
+    init {
+        LayoutInflater.from(getContext())
+            .inflate(R.layout.home_item_card, this)
```
This can be replaced by the View `inflate` method...

inflate(context, R.layout.home_item_card, this)

leonardoxh commented on app/src/main/java/fr/groggy/racecontrol/tv/ui/home/HomeItemPresenter.kt:
```diff
@@ -0,0 +1,31 @@
+        return ViewHolder(view)
+    }
+
+    override fun onBindViewHolder(viewHolder: ViewHolder?, item: Any?) {
```
I think the viewHolder is never nullable in this case... if you remove the `?` from the parameter does it work?

kleinMaggus replied:
Yeah, it did work :)

leonardoxh commented on app/src/main/java/fr/groggy/racecontrol/tv/ui/season/archive/SeasonArchiveFragment.kt:
```diff
@@ -13,28 +13,43 @@ import fr.groggy.racecontrol.tv.ui.season.browse.SeasonBrowseActivity
         val viewModel: SeasonArchiveViewModel by viewModels()
         val archives = viewModel.listArchive()
-        itemAdapter.setItems(archives, null)
+        val decades = archives.groupBy { archive: Archive -> (archive.year / 10) * 10 }
```
I'd omit the `archive: Archive ->` and simple use as `it`

leonardoxh commented on app/src/main/java/fr/groggy/racecontrol/tv/ui/season/archive/SeasonArchiveFragment.kt:
```diff
@@ -13,28 +13,43 @@ import fr.groggy.racecontrol.tv.ui.season.browse.SeasonBrowseActivity
+    }
 
+    private fun buildRowsAdapter() {
+        val archivesTitle = "%s - %s"
```
Can you move this to the strings resource file? As this is a UI related stuff, if for some reason we decide to add anything on future it is a single place to change

leonardoxh commented on app/src/main/res/layout/activity_home.xml:
```diff
@@ -0,0 +1,45 @@
+        android:name="fr.groggy.racecontrol.tv.ui.home.HomeFragment"
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:keepScreenOn="true" />
```
The `keepScreenOn` here can be removed, this basically do not allow the screen to sleep.

leonardoxh commented on app/src/main/java/fr/groggy/racecontrol/tv/ui/season/archive/SeasonArchiveFragment.kt:
```diff
@@ -37,18 +37,19 @@ class SeasonArchiveFragment: BrowseSupportFragment(), OnItemViewClickedListener
+            val entries = decade.value
+            val listRowAdapter = ArrayObjectAdapter(ArchivePresenter())
+            listRowAdapter.setItems(entries, null)
+            val title = getString(R.string.season_archives_title).format(
+                entries[0].year,
+                entries[entries.size - 1].year
+            )
```
replace with getString(id, args...)
